File: CS/XtraTabWithButtons/Tab/Page/XtraTabPageDescendant.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraTab;
using DevExpress.XtraTab.Buttons;
using System.ComponentModel;
using DevExpress.XtraEditors;
using DevExpress.XtraTab.ViewInfo;
using DevExpress.Utils.Design;

namespace XtraTabWithButtons
{
    public class XtraTabPageDescendant : XtraTabPage
    {
        CustomHeaderButtonCollection customHeaderButtonsCore;

        public event CustomHeaderButtonEventHandler CustomPageButtonClick;

        public XtraTabPageDescendant()
        {
            customHeaderButtonsCore = new CustomHeaderButtonCollection();
            CustomButtons.CollectionChanged += CustomButtons_CollectionChanged;
        }

        void CustomButtons_CollectionChanged(object sender, CollectionChangeEventArgs e)
        {
            if (!DesignMode || Site == null) return;
            EditorContextHelper.FireChanged(Site, this);
        }

        [Localizable(true), Category(CategoryName.Behavior), RefreshProperties(RefreshProperties.All)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public virtual CustomHeaderButtonCollection CustomButtons
        {
            get { return customHeaderButtonsCore; }
        }

        public void RaiseCustomPageButtonClick(TabButtonInfo button)
        {
            if (CustomPageButtonClick != null)
            {
                CustomHeaderButtonEventArgs args = new CustomHeaderButtonEventArgs(button.Button as CustomHeaderButton, this);
                CustomPageButtonClick(this, args);
            }
        }
    }
}

File: VB/XtraTabWithButtons/Tab/Page/XtraTabPageDescendant.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraTab
Imports DevExpress.XtraTab.Buttons
Imports System.ComponentModel
Imports DevExpress.XtraEditors
Imports DevExpress.XtraTab.ViewInfo
Imports DevExpress.Utils.Design

Namespace XtraTabWithButtons
	Public Class XtraTabPageDescendant
		Inherits XtraTabPage
		Private customHeaderButtonsCore As CustomHeaderButtonCollection

		Public Event CustomPageButtonClick As CustomHeaderButtonEventHandler

		Public Sub New()
			customHeaderButtonsCore = New CustomHeaderButtonCollection()
			AddHandler CustomButtons.CollectionChanged, AddressOf CustomButtons_CollectionChanged
		End Sub

		Private Sub CustomButtons_CollectionChanged(ByVal sender As Object, ByVal e As CollectionChangeEventArgs)
			If (Not DesignMode) OrElse Site Is Nothing Then
				Return
			End If
			EditorContextHelper.FireChanged(Site, Me)
		End Sub

		<Localizable(True), Category(CategoryName.Behavior), RefreshProperties(RefreshProperties.All), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Overridable ReadOnly Property CustomButtons() As CustomHeaderButtonCollection
			Get
				Return customHeaderButtonsCore
			End Get
		End Property

		Public Sub RaiseCustomPageButtonClick(ByVal button As TabButtonInfo)
			If CustomPageButtonClickEvent IsNot Nothing Then
				Dim args As New CustomHeaderButtonEventArgs(TryCast(button.Button, CustomHeaderButton), Me)
				RaiseEvent CustomPageButtonClick(Me, args)
			End If
		End Sub
	End Class
End Namespace
